functions/go/set-labels/transformer.go

```diff
@@ -0,0 +1,301 @@
+package main
+
+import (
+	"sort"
+
+	"github.com/GoogleContainerTools/kpt-functions-sdk/go/fn"
+)
+
+type FieldPath []string
+var (
+	selectorPath = FieldPath{"selector", "matchLabels"}
+	metaLabelsPath = FieldPath{"metadata", "labels"}
+	podSelectorPath = FieldPath{"podSelector", "matchLabels"}
+
+)
+
+
+var _ fn.Runner = &SetLabels{}
+
+// SetLabels supports the set-labels workflow, it uses Config to parse functionConfig, Transform to change the labels
+type SetLabels struct {
+	// labels is the desired labels
+	Labels map[string]string `json:"labels,omitempty"`
+	count int
+}
+
+// Transform updates the labels in the right path using GVK filter and other configurable fields
+func (p *SetLabels) Run(_ *fn.Context, fnConfig *fn.KubeObject, items fn.KubeObjects, results *fn.Results) bool {
+	objects := items.WhereNot(fn.IsLocalConfig)
+	if objects.Len() == 0 || objects[0] == nil {
+		results.Infof("no input resources")
+		return true
+	}
+	// A workaround since kpt creates a FunctionConfig placeholder if users don't provide the functionConfig.
+	if fn.EmptyFunctionConfig(fnConfig) {
+		return false
+	}
+	if len(p.Labels) == 0 {
+		results.Warningf("no `labels` arguments are given in FunctionConfig")
+		return true
```

Comment:
They have some slice differences:

Since `set-labels` requires a functionConfig, if users do not ever provide one, the run should fail, and should do action as "adding a functionconfig" (this is also why I need to distinguish the kpt fnConfig placeholder because it makes the fnConfig non-empty). But if the function config is provided, even if no labels (which is allowed), I cannot think of a reason to fail the run (so only give warnings). It's more like you did run the function but it's a no-op.

To give more context: when I myself use a function, in many times it passes with no-op or fails with irrelevant errors. Most of the time, it's because my functionconfig or some other setup is incorrect, but I cannot get efficient info from the results. So I think one big focus of the function is to help users to debug their config error.  Thus, it seems worth the effort to be very specific and crystal clear on the users messages.

Reply:
This is very good.